### PR TITLE
feat(python)!: Set default name for expression output of some functions

### PR DESCRIPTION
**Pull request by stinodego**
Partially addresses #8972

Changes:
* For the following functions, the name of the output expression is now set to a default, rather than set to the name left-most input expression.
  * `concat_list` -> `"list"`
  * `struct` -> `"struct"`
  * `concat_str` -> `"string"`
  * `format` -> `"string"`

This can break your workflow if you rely on the default name. Consider the following code:

```python
df = pl.DataFrame({"a": [1, 2], "b": [3, 4]})
result = df.with_columns(pl.struct("a", "b"))
print(result)
```

Result before:
```
shape: (2, 2)
┌───────────┬─────┐
│ a         ┆ b   │
│ ---       ┆ --- │
│ struct[2] ┆ i64 │
╞═══════════╪═════╡
│ {1,3}     ┆ 3   │
│ {2,4}     ┆ 4   │
└───────────┴─────┘
```

Result after:
```
shape: (2, 3)
┌─────┬─────┬───────────┐
│ a   ┆ b   ┆ struct    │
│ --- ┆ --- ┆ ---       │
│ i64 ┆ i64 ┆ struct[2] │
╞═════╪═════╪═══════════╡
│ 1   ┆ 3   ┆ {1,3}     │
│ 2   ┆ 4   ┆ {2,4}     │
└─────┴─────┴───────────┘
```

It is recommended to explicitly call `alias` on the expression if you want to overwrite the default name.

I think this should be mentioned explicitly in the `0.18.0` changelog. We can include the text above.

### CODE REVIEW

**Comment by ritchie46** [comment]
If we do this we shouldn't do this on the python side with an alias expression. This can lead to an alias call on evry group during aggregations, and for many expressions. 

If we change this, it should be set directly in the specific function that materializes them.

As an aside. The left-most rule for containers seemed to work pretty well IMO. Not a very strong opinion on my part, but it followed a consistent rule and had an ouput name that is linked to the input

File: polars/polars-lazy/polars-plan/src/dsl/functions.rs [hunk 1]
```diff
@@ -264,7 +264,8 @@ pub fn concat_str<E: AsRef<[Expr]>>(s: E, separator: &str) -> Expr {
             auto_explode: true,
             ..Default::default()
         },
-    }
+    };
+    expr.alias("string")
```
Comment (ritchie46) [inline polars/polars-lazy/polars-plan/src/dsl/functions.rs:268]: We should solve this in the `ConcatHorizontal` implementation.

File: polars/polars-lazy/polars-plan/src/dsl/functions.rs [hunk 2]
```diff
@@ -1301,6 +1303,7 @@ pub fn as_struct(exprs: &[Expr]) -> Expr {
         options.pass_name_to_apply = true;
         options
     })
+    .alias("struct")
```
Comment (ritchie46) [inline polars/polars-lazy/polars-plan/src/dsl/functions.rs:1306]: See earlier comment.

File: polars/polars-lazy/polars-plan/src/dsl/functions.rs [hunk 3]
```diff
@@ -298,21 +299,22 @@ pub fn format_str<E: AsRef<[Expr]>>(format: &str, args: E) -> PolarsResult<Expr>
...
         function: FunctionExpr::ListExpr(ListFunction::Concat),
         options: FunctionOptions {
             collect_groups: ApplyOptions::ApplyGroups,
             input_wildcard_expansion: true,
             fmt_str: "concat_list",
             ..Default::default()
         },
-    })
+    };
+    Ok(expr.alias("list"))
```
Comment (ritchie46) [inline polars/polars-lazy/polars-plan/src/dsl/functions.rs:317]: See earlier comment.

File: py-polars/polars/expr/list.py [hunk 4]
```diff
@@ -263,8 +263,9 @@ def concat(self, other: list[Expr | str] | Expr | str | Series | list[Any]) -> E
         other_list: list[Expr | str | Series]
         other_list = [other] if not isinstance(other, list) else copy.copy(other)  # type: ignore[arg-type]
 
-        other_list.insert(0, wrap_expr(self._pyexpr))
-        return F.concat_list(other_list)
+        original_expr = wrap_expr(self._pyexpr)
+        other_list.insert(0, original_expr)
+        return F.concat_list(other_list).alias(original_expr.meta.output_name())
```
Comment (ritchie46) [inline py-polars/polars/expr/list.py:268]: See earlier comment.

**Comment by stinodego** [comment]
Hm, this change isn't as obvious as it appeared at first.

> If we change this, it should be set directly in the specific function that materializes them.
 
I had it like that at first, e.g. did a rename in `concat_str` here:
https://github.com/pola-rs/polars/blob/e03183f55189b92a63128868f398f3d66dd93e80/polars/polars-core/src/functions.rs#L117

However, that also resulted in things like `pl.col('a') + pl.col('b')` getting renamed to `"string"`, as this uses the `concat_str` implementation. That is definitely undesirable. I guess we can add an optional `rename` argument, but that also doesn't feel great.

> As an aside. The left-most rule for containers seemed to work pretty well IMO. Not a very strong opinion on my part, but it followed a consistent rule and had an ouput name that is linked to the input

It's a tough one, to be honest. If you check the example in the PR description, I think that the new behaviour is much better. But I want to keep the behaviour that `pl.col('a') + pl.col('b')` results in the name `'a'`. But then `pl.concat_x(a, b)` and `a + b` have different behaviours, which I also don't really like.

And what do you mean by 'containers' exactly? Nested types like List and Struct? Or also Strings? It becomes finnicky - meanwhile `datetime(pl.col('a'))` will just rename the output to `"datetime"`, same with `date_range(pl.col('a'), ...)`.

Maybe this should be closed for now until we find a consistent naming strategy we like. Better not to break things unless we really like the new behaviour.

**Comment by stinodego** [comment]
Hey @mcrumiller,

I saw your suggestion but had to let it sink in and didn't get around to responding. I don't think it's a good idea. 

A global cache like that is not transparent to the user: for a certain operation, they won't know whether they will be getting `col_0` or `col_26`. I should be able to look at the code and know what will happen. It should not depend on what happens elsewhere in the code base.

I have previously suggested allowing nameless expressions (e.g. `None`). I think that would be the best way to deal with default name conflicts. But that has a lot of impact and we have so far decided not to go for it. As long as users are conscious about naming their stuff, and we have good error messages on naming conflicts, I think it's OK.

Your suggestion also doesn't address the issue raised by Ritchie that we might want to keep the left-most expression input name. Do we keep that name, or do we go for something else (a default, or a string cache)?

**Comment by mcrumiller** [comment]
@stinodego all good points. I suppose there are downsides and upsides to any approach, and recommending that users be explicit resolves the most downsides.

**Pull request closed without merging**